File: medium/994-Rotting-Oranges/BFS.kt
class Solution {
    fun orangesRotting(grid: Array<IntArray>): Int {
        val m = grid.size
        val n = grid[0].size
        
        val queue = ArrayDeque<Pair<Int, Int>>()
        val freshes = HashSet<Pair<Int, Int>>()
        
        for (i in 0 until m) {
            for (j in 0 until n) {
                if (grid[i][j] == 1) freshes.add(i to j)
                else if (grid[i][j] == 2) queue.offer(i to j)
            }
        }
        
        var counter = 0
        val directions = listOf(0 to 1, 1 to 0, -1 to 0, 0 to -1)
        while(!queue.isEmpty() && !freshes.isEmpty()) {
            repeat(queue.size) {
                val (i, j) = queue.poll()
                for ((di, dj) in directions) {
                    val adjacent = (i + di to j + dj)
                    if (adjacent in freshes) {
                        freshes.remove(adjacent)
                        queue.offer(adjacent)
                    }
                }
            }
            ++counter
        }
        
        return if (freshes.isEmpty()) counter else -1
    }
}

File: medium/994-Rotting-Oranges/BFS.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])

        queue = deque()
        freshes = set()
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    freshes.add((i, j))
                elif grid[i][j] == 2:
                    queue.append((i, j))

        counter = 0
        while queue and freshes:
            n_iter = len(queue)
            for _ in range(n_iter):
                i, j = queue.popleft()
                for di, dj in [(0, 1), (1, 0), (-1, 0), (0, -1)]:
                    adjacent = (i + di, j + dj)
                    if adjacent in freshes:
                        freshes.remove(adjacent)
                        queue.append(adjacent)
            counter += 1

        return counter if len(freshes) == 0 else -1
